### PR TITLE
Feature/signed fetch

### PR DESCRIPTION
**Pull request by jdahlke**
- [x] added signedFetch for HMAC authentication
- [x] different **browser** and **server** clients for service tokens
- [x] remove from jasmine to mocha + chai

### CODE REVIEW

File: src/utils/simple_fetch.js [hunk 1]
```diff
@@ -48,7 +51,9 @@ function error (e) {
     return resolve({
```
Comment (Skudo) [inline src/utils/simple_fetch.js]: Why do we `resolve` a promise, when an error actually occurred? I feel like this behaviour is causing weird things later on, like having to actually look into the actual response to determine whether something went wrong in the first place.

I feel like that's the only reason why there's a `response.ok` attribute.

File: src/index.js [hunk 2]
```diff
@@ -2,10 +2,13 @@
...
-import ServiceToken from 'src/shark/service_token'
 
-import ClientError from 'src/shark/client_error'
-import ServerError from 'src/shark/server_error'
+import BrowserServiceTokenClient from 'src/shark/service_token/node_client'
+import NodeServiceTokenClient from 'src/shark/service_token/browser_client'
+
+import SharkError from 'src/shark/error'
+
+const ServiceToken = Config.nodeProcess ? NodeServiceTokenClient : BrowserServiceTokenClient
```
Comment (Skudo) [inline src/index.js]: I don't have a really good, useful way to do the entire "We need two different clients for _mostly_ the same thing better", but maybe we should have a `service_token_client.js` that acts as kind of a factory by hiding the decision which client to use.

File: src/shark/service_token/node_client.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    return this.__request(url, requestOptions)
+  }
+
+  /**
+   * Verifies if a service token is still valid.
+   *
+   * @return {Promise} the fetch promise
+   */
+  verifyServiceToken (params, options = {}) {
+    const url = `${this.baseUrl}/api/users/authenticate?service_token=${params.serviceToken}`
```
Comment (Skudo) [inline src/shark/service_token/node_client.js]: That weird construct with the `urls` object containing functions to return URLs for each function in my original implementation was so you could override the actual URL more easily. Whilst I guess _that_ much "power" might have been a bit too much, I think being able to specify additional query string parameters such as `include` might be relevant at some point. Especially if we need do include `permissions` in some, but not all objects.

Alternatively, `verifyServiceToken` could _always_ have `include=permissions` in the URL. Given that some actions like creating notifications also care about authorisation, that's necessary, unless we allow for a way to modify the endpoint URL.

File: src/shark/service_token/node_client.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      secretKey: this.secretKey,
+      digest: this.digest,
+      headers: headers
+    })
+
+    return signedFetch(url, requestOptions).then(json => {
+      return deserializer.deserialize(json)
+    }).then(serviceToken => {
+      return serviceToken
+    })
```
Comment (Skudo) [inline src/shark/service_token/node_client.js]: Couldn't you just drop this second `.then`... 🤔

(Yes, you probably can.)

File: src/index.js [hunk 5]
```diff
@@ -2,10 +2,13 @@
 
 import Config from 'src/shark/config'
 import Client from 'src/shark/client'
-import ServiceToken from 'src/shark/service_token'
 
-import ClientError from 'src/shark/client_error'
-import ServerError from 'src/shark/server_error'
+import BrowserServiceTokenClient from 'src/shark/service_token/node_client'
+import NodeServiceTokenClient from 'src/shark/service_token/browser_client'
```
[the remark this answers is not in the document]
Comment (natalka) [reply src/index.js]: exactly, its very confusing... 🤔

File: src/utils/simple_fetch.js (hunk 1 above)
Comment (jdahlke) [reply src/utils/simple_fetch.js]: I need one format (the response format) to fit the later promise pipeline, which decides whether to reject or resolve the `fetch` promise. I don't want the client to use different error formats for connection error and e.g. ResourceNotFound.

File: src/index.js (hunk 2 above)
Comment (jdahlke) [reply src/index.js]: I kind of thought this is the factory.
I can put these lines in an extra file, but I don't think that would change anything at all.

File: src/shark/service_token/node_client.js (hunk 3 above)
Comment (jdahlke) [reply src/shark/service_token/node_client.js]: Actually it should be in the params:
`params = { serviceToken: '0123456789', include: 'permissions' }`. 

I just forgot to add the params to the query string.

**Pull request merged**